library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -1,65 +1,216 @@
+  m LLVM.Operand
+mkMain t@(ErasedAnn.Ann usage ty t') = do
+  let types = map typeToLLVM (functionTy ty)
+      returnType = P.last types
+      paramTypes = init types
+      paramNames = repeat "arg"
+      params = zip paramTypes (map mkParameterName paramNames)
+  LLVM.function "main" params returnType $ \args -> do
+    out <- case t' of
+      ErasedAnn.LamM
+        { ErasedAnn.capture,
+          ErasedAnn.arguments,
+          ErasedAnn.body
+        } -> do
+          let env = Map.fromList $ zip paramNames args
+              callArgs = zip args (repeat []) -- No arg attributes.
+          funname <- mkLam env ty body arguments capture
+          LLVM.call (globalRef (typeToLLVM ty) funname) callArgs
+      _ -> compileTerm mempty t
+    LLVM.ret out
+
+-- | Compile a term to its equivalent LLVM code.
+-- TODO: implement other constructors of ErasedAnn.Term
 compileTerm ::
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of Juvix variables to LLVM function arguments.
+  Env ->
+  -- | The term to compile.
   ErasedAnn.AnnTerm PrimTy RawPrimVal ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-compileTerm (ErasedAnn.Ann usage ty t) = case t of
-  ErasedAnn.Prim t' -> mkPrim t
-  ErasedAnn.AppM f xs -> mkApp (ErasedAnn.term f) ty xs
+  m LLVM.Operand
+compileTerm env (ErasedAnn.Ann usage ty t) = case t of
+  ErasedAnn.Var symbol -> case Map.lookup symbol env of
+    Nothing ->
+      P.error $ "Variable not found: " <> show symbol <> " in " <> show env -- TODO improve error message.
+    Just var -> return var
+  ErasedAnn.Prim t' -> mkPrim t ty
+  ErasedAnn.AppM f xs -> mkApp env f ty xs
 
-mkPrim ::
-  ErasedAnn.Term PrimTy RawPrimVal ->
-  LLVM.IRBuilderT LLVM.ModuleBuilder LLVM.Operand
-mkPrim (ErasedAnn.Prim prim) = case prim of
-  LitInt i -> return $ LLVM.int8 i -- TODO deal with the type correctly.
+-- | Write an LLVM function definition based on a the given lambda abstraction.
+-- The function returns the name of the create function.
+mkLam ::
+  (LLVM.MonadIRBuilder m, LLVM.MonadModuleBuilder m) =>
+  -- | Environment of Juvix variables to LLVM function arguments.
+  Env ->
+  -- | The type of the lambda abstraction.
+  ErasedAnn.Type PrimTy ->
+  -- | The body of the lambda abstraction.
+  ErasedAnn.AnnTerm PrimTy RawPrimVal ->
+  -- | List of parameter names.
+  [NameSymbol.T] ->
+  -- | List of captures variables (free variables for the body).
+  [NameSymbol.T] ->
+  m LLVM.Name
+mkLam env ty body args capt = do
+  funname <- LLVM.freshName "lam"
+  let returnTy = typeToLLVM $ ErasedAnn.type' body
+  params <- mapM mkParam (zip args (functionTy ty))
+  LLVM.function funname params returnTy $ \refs -> do
+    let env' = env `Map.union` Map.fromList (zip args refs)
+    body' <- compileTerm env' body
+    LLVM.ret body'
+  return funname
+  where
+    -- Given a Juvix name and type, construct an LLVM function parameter
+    -- definition.
+    mkParam ::
```

Comment:
this should be a top-level function perhaps?

Reply:
I am not particularly happy with it either, there is also some overlap with mkParameterName.

Reply:
Can we change it to a top-level function?

Reply:
I have addressed this and replaced it by the clearer `mkParameter` global function.